### PR TITLE
mobile/ci: Switch to Linux for running Android apps

### PR DESCRIPTION
**Pull request by fredyw**
This PR switches the Android builds to run on Linux (with larger runners) since they have hardware acceleration support, which should make running the Android app tests much faster.

Risk Level: low
Testing: CI
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: mobile/ci/linux_ci_setup.sh
Comment (phlax) [inline mobile/ci/linux_ci_setup.sh:5]: if you use the mobile build image this shouldnt be necessary - i think it was why it was added

Comment (phlax) [reply mobile/ci/linux_ci_setup.sh:5]: my misgiving is that it looks like another install to me, when i think we should be going in the opposite direction - but i dont know the specifics

Comment (fredyw) [reply mobile/ci/linux_ci_setup.sh:5]: This is for running the emulator. Running the emulator inside Docker, although not impossible, requires a lot of work, such as installing the GUI libraries, etc. We can explore that option in the future once we have a stable CI. For now the goal of this PR is simply to test switching to Linux with a large runner.

File: .github/workflows/mobile-android_build.yml
Comment (RyanTheOptimist) [inline .github/workflows/mobile-android_build.yml:47]: Out of curiosity, why this change?

File: .github/workflows/mobile-android_build.yml
Comment (RyanTheOptimist) [inline .github/workflows/mobile-android_build.yml:61]: I know this is the current name of this runner, but I wonder if we should cook up a name that at least has linux in the name? But that's out of scope for this PR, obviously.

File: .github/workflows/mobile-android_build.yml
Comment (RyanTheOptimist) [inline .github/workflows/mobile-android_build.yml:220]: nit: Do we need quotes here?

Would it make sense to put "${ANDROID_HOME}/platform-tools/" into PATH so we don't have to repeat the full path to adb.

**Review by RyanTheOptimist — COMMENTED** [review]
This is amazing!

File: .github/workflows/mobile-android_build.yml
Comment (fredyw) [reply .github/workflows/mobile-android_build.yml:47]: That's because `remote-ci` has [--remote_download_minimal](https://github.com/envoyproxy/envoy/blob/e9366f2b34be537b1b30ed17a4fd75b4715b5990/.bazelrc#L391), which means the Bazel output will not be available.

File: .github/workflows/mobile-android_build.yml
Comment (fredyw) [reply .github/workflows/mobile-android_build.yml:61]: +1. I also think we should call this something like `ubuntu-22.04-large` or something like that.

File: .github/workflows/mobile-android_build.yml
Comment (fredyw) [reply .github/workflows/mobile-android_build.yml]: We don't need the full quotes, but we definitely need to quote a variable, i.e. `"${ANDROID_HOME}"`. It's a good practice to do that. See https://www.shellcheck.net/wiki/SC2086

As for updating the `PATH`, typically we don't want to do that CI. Being explicit is always better IMO.

I updated the PR to only quote `"${ANDROID_HOME}"`, which is also what I typically do.

Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml]: Good point!

**Review by RyanTheOptimist — APPROVED** [review]

**Comment by RyanTheOptimist** [comment]
LGTM, but please wait for approval from @phlax, too.

**Review by abeyad — APPROVED** [review]

File: .github/workflows/mobile-android_build.yml
Comment (phlax) [reply .github/workflows/mobile-android_build.yml:47]: to expand on this, when i last worked on the bazel flags i implemented a pattern - more for Envoy, but also for mobile

Anything that ends `remote-ci` does not download - in the case of envoy the equivalent `--config=remote` + `--config=ci` are the same except that it doesnt enforce a download style

File: .github/workflows/mobile-android_build.yml
Comment (phlax) [reply .github/workflows/mobile-android_build.yml:61]: not sure its mobile specific

envoy-x64-$SIZE would be our current convention as used in azp

Comment (phlax) [reply .github/workflows/mobile-android_build.yml:61]: the envoy prefix is because the pools are shared (subject to restrictions etc) across the org so if any other projects require separate runners etc (eg nighthawk) then we know which is which

i would suggest fixing this before landing - its a pita changing the name once PRs are being raised against it

File: .github/workflows/mobile-android_build.yml
Comment (phlax) [inline .github/workflows/mobile-android_build.yml:76]: i think this and the adb install stuff belong in a github action

i can follow up to add and shift the code

**Review by phlax — APPROVED** [review]
lgtm, thanks @fredyw 

i would prefer we fix the pool name before we land - i can help with that

File: .github/workflows/mobile-android_build.yml
Comment (fredyw) [reply .github/workflows/mobile-android_build.yml:76]: Thanks.

**Review by phlax — APPROVED** [review]
lgtm, thanks @fredyw

File: .github/workflows/mobile-android_build.yml
Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml:76]: I totally trust your judgement here, @phlax. Can you explain more about what makes sense in the workflow and what make sense in an action? I have too little experience to have a clue :)

File: .github/workflows/mobile-android_build.yml
Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml]: I see we settled on `envoy-x64-small` for this runner. I'm surprised that we don't have the OS in the name. This makes it harder for us to tell at a glance which CI jobs are using which platforms (and the info on the runners is hidden in github UI that we don't have access to). Can we revisit this naming scheme? Alternatively, maybe we could publish info on our runners in some doc somewhere?

File: .github/workflows/mobile-android_build.yml
Comment (phlax) [reply .github/workflows/mobile-android_build.yml:76]: well, kinda forming strategies as we speak - but probs we dont want any bash etc in workflows - split those out to actions for reusablitiy and isolation and use workflows for controlling _workflows_

Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml:76]: Ah, that makes sense.

File: .github/workflows/mobile-android_build.yml
Comment (phlax) [reply .github/workflows/mobile-android_build.yml]: >  I'm surprised that we don't have the OS in the name.

no point really - all of the ms hosted linux are ubuntu - and similar with azp and our self-hosted

> This makes it harder for us to tell ...

assume linux/ubuntu and all will be clear

> Alternatively, maybe we could publish info on our runners in some doc somewhere?

happy to review - but its something that changes over time with necessity/dev so it would add maintenance - the ci yaml source is kinda the authority here

perhaps we can add a high level - this is _how_ they are named doc

Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml]: > > I'm surprised that we don't have the OS in the name.
> 
> no point really - all of the ms hosted linux are ubuntu - and similar with azp and our self-hosted

So some runners are macOS, some are linux and (pending the Windows work) some will be windows. This seems like an important detail that would be easy to encode in the runner name without causing any problems. Looking at the "runs-on" fields in our .yml files today we have:
```
     14     runs-on: ${{ needs.env.outputs.agent_ubuntu }}
     18     runs-on: macos-12
      5     runs-on: ubuntu-20.04
     21     runs-on: ubuntu-22.04
```

All of them seem to have the OS in the name, and this seems really useful.
 
> > This makes it harder for us to tell ...
> 
> assume linux/ubuntu and all will be clear

Why assume when we could be explicit? I think it would be better to have a naming scheme that includes the OS.

> > Alternatively, maybe we could publish info on our runners in some doc somewhere?
> 
> happy to review - but its something that changes over time with necessity/dev so it would add maintenance 

If we have a naming scheme which includes the relevant info then there is no need for any docs, of course. That seems like the least burden.

> - the ci yaml source is kinda the authority here

Oh, is there a yaml file I could look at which would tell me that `envoy-x64-small` is linux? If so, that's great.

> perhaps we can add a high level - this is _how_ they are named doc

That would be great too.

Comment (phlax) [reply .github/workflows/mobile-android_build.yml]: > So some runners are macOS, some are linux and (pending the Windows work) some will be windows. This seems like an important detail that would be easy to encode in the runner name without causing any problems. Looking at the "runs-on" fields in our .yml files today we have:

thas kinda the point in a way - macos version **determines** the arch, windows is only x64 

linux is the only one with variants, and they are **all** ubuntu

this is how we came to the naming scheme that we have used so far (cc @mum4k ) and tbh i think it makes sense - no need to add redundant info

> Oh, is there a yaml file I could look at which would tell me that envoy-x64-small is linux? If so, that's great.

no but you can do `git grep runs-on` to find out what is currently used

Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml]: > > So some runners are macOS, some are linux and (pending the Windows work) some will be windows. This seems like an important detail that would be easy to encode in the runner name without causing any problems. Looking at the "runs-on" fields in our .yml files today we have:
> 
> thas kinda the point in a way - macos version **determines** the arch, windows is only x64
> 
> linux is the only one with variants, and they are **all** ubuntu
> 
> this is how we came to the naming scheme that we have used so far (cc @mum4k ) and tbh i think it makes sense - no need to add redundant info
> 
> > Oh, is there a yaml file I could look at which would tell me that envoy-x64-small is linux? If so, that's great.
> 
> no but you can do `git grep runs-on` to find out what is currently used

Yes, that's what I did and you'll notice that all the existing runners (except the new one we're talking about here) have the OS in the name:
* macos-12
* ubuntu-20.04
* ubuntu-22.04
The only runner that does NOT have the OS name is `envoy-x64-small`. I'm suggesting that we stick `linux` or `ubunutu` (or `ubuntu-22.04`) in the name. I don't understand what the upside is to NOT having the OS in the name, especially since the GH runner names all do.

Comment (phlax) [reply .github/workflows/mobile-android_build.yml]: that is the ms-hosted naming - we made up our own when managing self-hosted pools for various reasons

apart from anything _not_ doing that emphasizes these are not the standard vms

Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml]: I don't think I understand. The current runners in the yaml file have OS in the name. For at least mobile developers, having the OS in the runner name makes it easier to reason about our CI jobs. The new runner does not have OS in the name. I think we should rename it to include the OS in the name (`envoy-x64-linux-small`, `envoy-x64-ubuntu-small`, `envoy-x64-ubuntu-22.04-4cpu-16gb`. Can we just do that? Or is there some reason it would be *bad* to have the OS in the name? If there is, I don't think I understand what it is.

Comment (phlax) [reply .github/workflows/mobile-android_build.yml]: > The current runners in the yaml file have OS in the name

only the ms-hosted ones

i think its generally useful to be able to discriminate them in that way

Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml]: > > The current runners in the yaml file have OS in the name
> 
> only the ms-hosted ones

Are there any runners in the current yaml file `runs-on:` which do not have OS in the name?
 
> i think its generally useful to be able to discriminate them in that way

Do I understand correctly that you think we should NOT put OS in the name of the github runners we create because the MS runners have the OS name and so if a runner does not have the name we know it's a custom GH runner? I'd certainly be happy to see some token in the name to discriminate between MS-hosted runners and GH-custom runners (perhaps "-ms" and "-gh", but I'm game for anything). But the lack of OS in the name is not compelling.

But maybe I misunderstanding your point.

Comment (phlax) [reply .github/workflows/mobile-android_build.yml]: > Are there any runners in the current yaml file runs-on: which do not have OS in the name?

yes - all of the "custom" ones

> Do I understand correctly that you think we should NOT put OS in the name of the github runners we create because the MS runners have the OS name and so if a runner does not have the name we know it's a custom GH runner?

yes and other than in the case of mac or win - the platform is superfluous its **always** linux and ubuntu

Comment (phlax) [reply .github/workflows/mobile-android_build.yml]: and in the case of mac or win the arch is determined

so in those cases - as you can see with current windows image - i have named the first pool `envoy-win19-medium` 

as that has **all** of the salient information with the caveat that we could/should define somewhere what small/medium/large/etc means (and i guess where thay can be used)

Comment (phlax) [reply .github/workflows/mobile-android_build.yml]: ... and `envoy-win22-medium` (for ref)

Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml]: > > Are there any runners in the current yaml file runs-on: which do not have OS in the name?
> 
> yes - all of the "custom" ones

What's an example of what you're talking about? I don't seem to be able to find them:

```
% git grep -h 'runs-on:'  | sort | uniq -c | sort -nr       
     21     runs-on: ubuntu-22.04
     14     runs-on: macos-12
     14     runs-on: ${{ needs.env.outputs.agent_ubuntu }}
      5     runs-on: ubuntu-20.04
      4     runs-on: envoy-x64-small
      1       runs-on: envoy-x64-small
      1     runs-on: ${{ inputs.runs-on }}
      1       runs-on:
```
 
> > Do I understand correctly that you think we should NOT put OS in the name of the github runners we create because the MS runners have the OS name and so if a runner does not have the name we know it's a custom GH runner?
> 
> yes and other than in the case of mac or win - the platform is superfluous its **always** linux and ubuntu

It's clearly useful to know if the runner is mac, win or linux. I don't think there is any reason to omit this information just for custom github linux runners.

Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml]: > `envoy-win22-medium`

This seems like a great name. Let's just add linux or ubunutu to the new runner to match that pattern (and/or remove x64. I don't have any opinion about that)

Comment (phlax) [reply .github/workflows/mobile-android_build.yml]: i dont agree for the reasons explained

Comment (RyanTheOptimist) [reply .github/workflows/mobile-android_build.yml]: @alyssawilk when you're back, can you please tie-break here.

Comment (phlax) [reply .github/workflows/mobile-android_build.yml]: you win - happy to review

**Pull request merged**
